{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref = \"C:/Users/Flora/Desktop/M2 Systèmes Complexes/Stage/Bioinfo/ref/\"\n",
    "data = \"C:/Users/Flora/Desktop/M2 Systèmes Complexes/Stage/Bioinfo/LTEE-clone-curated/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_positions(IS_name):\n",
    "    '''Generates a file listing the position of insertions that occured at some point in at least\n",
    "    one of the clones of one of the populations (only one occurrence per position, even if it occured\n",
    "    in different populations).'''\n",
    "    \n",
    "    from os import listdir\n",
    "    from os.path import isfile, join\n",
    "    onlyfiles = [f for f in listdir(data) if isfile(join(data, f))] #listing all file names for the .gd files\n",
    "    files = [] # creating a list that will first contain these names to be called, and later on open .gd files\n",
    "    file_names = [] # list that will contain names as strings (kept as such for all the analysis)\n",
    "    readlines_names = [] #list that will contain lists with the lines of each file (obtained via readlines())\n",
    "    for i in onlyfiles: #fixing an error in the list of files (I don't know why it occurs) and filling the previously defined lists\n",
    "        if len(i) < 30:\n",
    "            files.append(i)\n",
    "            file_names.append(i)\n",
    "            readlines_names.append(i)\n",
    "    \n",
    "    \n",
    "    for i in range(len(files)):\n",
    "        files[i] = open(data+files[i],\"r\") #the list files becomes a list of open files\n",
    "        readlines_names[i] = files[i].readlines() # the list readlines_names becomes a list of lists, with each list containing \n",
    "        #all the lines from one file\n",
    "        \n",
    "    insertion_positions = {} # Creating a dictionary to hold the positions of the insertions\n",
    "    for i in range(len(file_names)): # for each file\n",
    "        for line in readlines_names[i]: #we look at all the lines in the corresponding readlines_names entry\n",
    "            line = line.split()\n",
    "            if 'MOB' in line and IS_name in line: #if the line corresponds to a mutation linked to the IS\n",
    "                insertion_positions[line[4]] = [line[4], line[6]] #we add an entry for this IS position to our dictionary\n",
    "                # and save the orientation as well\n",
    "    \n",
    "    return insertion_positions\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def interrupted_genes_proportion(insertion_positions):\n",
    "    '''Uses the list generated by extract_positions() to extract the type of sequence in which insertions occured.\n",
    "    Returns the proportion of interrupted sequences that are genes.'''\n",
    "    \n",
    "    ref_file = open(ref+'REL606.gff3')\n",
    "    ref_readlines = ref_file.readlines()\n",
    "    ref_readlines = ref_readlines[5:8897] # we don't need the 4 first lines or the full fasta sequence\n",
    "    \n",
    "    number_insertions_genes = 0\n",
    "    number_insertions_else = 0\n",
    "    \n",
    "    \n",
    "    #NB: there is a problem in this part, somehow some insertions get counted twice (at the end I have a total of\n",
    "    # 292 insertions instead of 270 for IS150) and I can't figure out the problem\n",
    "    for position in list(set(insertion_positions)):\n",
    "        line_found = False\n",
    "        for line_ref in ref_readlines:\n",
    "            if 'CDS' not in line_ref.split()[2]:\n",
    "                if int(position) >= int(line_ref.split()[3]) and int(position) <= int(line_ref.split()[4]):\n",
    "                    line_found = True\n",
    "                    if line_ref.split()[2] == 'gene':\n",
    "                        number_insertions_genes += 1\n",
    "                    else:\n",
    "                        number_insertions_else += 1\n",
    "        if not line_found:\n",
    "            number_insertions_else += 1\n",
    "            \n",
    "    print('total number of insertions: ', number_insertions_genes + number_insertions_else)\n",
    "    return number_insertions_genes/(number_insertions_genes + number_insertions_else)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_gene_proportion():\n",
    "    '''Returns the proportion of coding sequences in the genome of REL606.'''\n",
    "    \n",
    "    ref_file = open(ref+'REL606.gff3')\n",
    "    ref_readlines = ref_file.readlines()\n",
    "    ref_readlines = ref_readlines[5:8897] # we don't need the 4 first lines or the full fasta sequence\n",
    "    \n",
    "    list_lengths = []\n",
    "    list_types = []\n",
    "\n",
    "        \n",
    "    for line in ref_readlines:\n",
    "        line = line.split()\n",
    "        if 'CDS' in line[2]:\n",
    "            sequence_length = int(line[4]) - int(line[3]) # length of the segment described by the line\n",
    "            sequence_type = line[2]\n",
    "            list_lengths.append(sequence_length)\n",
    "            list_types.append(sequence_type)\n",
    "    \n",
    "    total_length = sum(list_lengths)\n",
    "    proportion_genes = 0\n",
    "    for i in range(len(list_types)):\n",
    "        if list_types[i] == 'gene':\n",
    "            proportion_genes += list_lengths[i]/total_length\n",
    "        \n",
    "    print('total number of genes: ', len(list_types))\n",
    "    return proportion_genes\n",
    "            \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total number of insertions:  292\n",
      "total number of genes:  4564\n",
      "0.7636986301369864\n",
      "0.9362800547629626\n"
     ]
    }
   ],
   "source": [
    "insertion_positions = extract_positions('IS150')\n",
    "interrupted_genes_prop = interrupted_genes_proportion(insertion_positions)\n",
    "proportion_genes_genome = extract_gene_proportion()\n",
    "print(interrupted_genes_prop)\n",
    "print(proportion_genes_genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total number of insertions:  22\n",
      "total number of genes:  4564\n",
      "0.6818181818181818\n",
      "0.9362800547629626\n"
     ]
    }
   ],
   "source": [
    "insertion_positions = extract_positions('IS186')\n",
    "interrupted_genes_prop = interrupted_genes_proportion(insertion_positions)\n",
    "proportion_genes_genome = extract_gene_proportion()\n",
    "print(interrupted_genes_prop)\n",
    "print(proportion_genes_genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy.stats as stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7.507261592935916e-21"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oddsratio, pvalue = stats.fisher_exact([[0.7636986301369864*292, 0.23630136986301364*292],\n",
    "                                      [0.9362800547629626*4729599, 0.06371994523703739*4729599]])\n",
    "pvalue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Lg=4729599"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.00022331528047529078"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oddsratio, pvalue = stats.fisher_exact([[0.6818181818181818*22, 0.31818181818181823*22],\n",
    "                                      [0.9362800547629626*4729599, 0.06371994523703739*4729599]])\n",
    "pvalue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.31818181818181823"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
